Add private API documentation

chrysle opened the pull request:
<!--- Describe the changes here. --->

Closes #1999. Thanks for providing the example, @webknjaz!

##### Contributor checklist

- [ ] Included tests for the changes.
- [ ] PR title is short, clear, and ready to be included in the user-facing changelog.

##### Maintainer checklist

- [x] Verified one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

webknjaz commented on docs/conf.py:
```diff
@@ -36,7 +36,7 @@
 # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
 # ones.
-extensions = ["myst_parser", "sphinxcontrib.programoutput"]
+extensions = ["myst_parser", "sphinxcontrib.programoutput", "sphinxcontrib.apidoc"]
```
Let's keep them sorted
```suggestion
extensions = [
    "myst_parser",
    "sphinxcontrib.apidoc",
    "sphinxcontrib.programoutput",
]
```

webknjaz commented on docs/requirements.txt:
```diff
@@ -2,7 +2,7 @@
 # This file is autogenerated by pip-compile with Python 3.11
 # by the following command:
 #
-#    pip-compile --allow-unsafe --strip-extras docs/requirements.in
```
It's best to use the same working dir as dependabot to avoid this header being updated back and forth all the time...

chrysle replied:
Though, in the last bot PR (#2043), things worked as expected?

webknjaz replied:
Ah, interesting. Anyways, please don't mix infra changes with the functional ones. We don't have an ability/infra to make two separate changelog entries per PR. With atomic PRs, this wouldn't be a problem.

webknjaz replied:
Marking this as resolved to unblock.

webknjaz commented on docs/requirements.in:
```diff
@@ -3,3 +3,4 @@ myst-parser
 setuptools-scm
 sphinx
 sphinxcontrib-programoutput
+sphinxcontrib-apidoc
```
```suggestion
sphinxcontrib-apidoc
sphinxcontrib-programoutput
```

webknjaz commented on piptools/resolver.py:
```diff
@@ -161,7 +161,9 @@ def resolve(self, max_rounds: int) -> set[InstallRequirement]:
     ) -> dict[InstallRequirement, set[str]]:
-        """Find acceptable hashes for all of the given ``InstallRequirement``s."""
+        """
+        Find acceptable hashes for all of the given ``InstallRequirement``'s.
```
This change is incorrect. It's not `'s` but just a plural `s`.

chrysle replied:
I wouldn't have added that if Sphinx hadn't given me this strange error:

```
/home/user/repos/github/pip-tools/.tox/build-docs/lib/python3.9/site-packages/piptools/resolver.py:docstring of piptools.resolver.BaseResolver.resolve_hashes:1: WARNING: Inline literal start-string without end-string.
```

webknjaz replied:
Either way, the grammar is broken. RST parser can be tricked into merging parts by adding an escaped whitespace:
```suggestion
        Find acceptable hashes for all of the given ``InstallRequirement``\ s.
```

If that doesn't work, the only thing left is rephrasing. There are other hacks, I think, but they are rather complicated and not worth it.

webknjaz commented on piptools/sync.py:
```diff
@@ -44,12 +44,12 @@ def dependency_tree(
-    Calculate the dependency tree for the package `root_key` and return
-    a collection of all its dependencies.  Uses a DFS traversal algorithm.
+    Calculate the dependency tree for the package ``root_key`` and return
+    a collection of all its dependencies. Uses a DFS traversal algorithm.
```
Many people have a habit of placing two spaces in between sentences. It's an old typographic practice. I think it's also mentioned in some PEPs. Since we don't enforce one way or the other (and RST is fine with either, I think), I wouldn't attempt reformatting within this PR (maybe ever).

But talking about PEP 257-compliant docstrings, they should have exactly one title/summary sentence on a separated line. That's something that would be related.

webknjaz commented:
I'd rename the PR title from "API" to "private API" for clarity.

webknjaz commented on tox.ini:
```diff
@@ -52,8 +52,9 @@ skip_install = true
 description = compile requirements for the documentation
+changedir = {[testenv:build-docs]changedir}
 commands_pre =
-commands = python -m piptools compile --strip-extras --allow-unsafe --quiet docs/requirements.in {posargs}
```
I'd leave this out of this PR. It's not related to documenting things.

chrysle commented:
I've pushed a commit following a valiant attempt to format `pip-tools`'s docstrings properly.

webknjaz commented on docs/conf.py:
```diff
@@ -58,3 +62,27 @@
 
 suppress_warnings = ["myst.xref_missing"]
+
+nitpick_ignore_regex = [
```
Let's put this right under `nitpicky` above.

webknjaz replied:
@chrysle looks like you missed this

webknjaz commented on docs/conf.py:
```diff
@@ -58,3 +62,27 @@
+    "--implicit-namespaces",
+    "--private",  # include “_private” modules
+]
+apidoc_module_dir = "../piptools"
+apidoc_module_first = False
```
Sorting
```suggestion
apidoc_module_first = False
apidoc_module_dir = "../piptools"
```

webknjaz commented on piptools/resolver.py:
```diff
@@ -182,6 +184,12 @@ def _filter_out_unsafe_constraints(
 
 class LegacyResolver(BaseResolver):
+    """
+    Resolve a given set of constraints (a collection of
```
It's a class, not a callable function. Classes represent static things/data. They don't do anything on their own. This is why, unlike functions/methods, their docstrings shouldn't start with a verb, because they don't perform actions, they just organize stuff. I usually recommend just stating what it is/represents in the docstring title line.

webknjaz commented on piptools/resolver.py:
```diff
@@ -193,11 +201,6 @@ def __init__(
         unsafe_packages: set[str] | None = None,
     ) -> None:
-        """
-        This class resolves a given set of constraints (a collection of
```
A docstring here may still be useful. It could describe what this initializer method does.

webknjaz commented on piptools/resolver.py:
```diff
@@ -416,6 +423,8 @@ def _iter_dependencies(
         self, ireq: InstallRequirement
     ) -> Iterator[InstallRequirement]:
         """
+        Collect all secondary dependencies for an ireq.
```
I usually go for "emit" in generator functions' docstrings. I feel like this word represents better that it can produce/stream more data, than just a single `return` instruction does.
```suggestion
        Emit all secondary dependencies for an ireq.
```

webknjaz commented on piptools/resolver.py:
```diff
@@ -629,7 +640,9 @@ def _do_resolve(
-        Return true on successful resolution, otherwise remove problematic
+        Actual resolution process.
+
+        Return True on successful resolution, otherwise remove problematic
```
Things like this are linkable
```suggestion
        Return :py:data:`True` on successful resolution, otherwise remove problematic
```

webknjaz commented on piptools/resolver.py:
```diff
@@ -629,7 +640,9 @@ def _do_resolve(
     ) -> bool:
         """
-        Return true on successful resolution, otherwise remove problematic
+        Actual resolution process.
```
This should still start with a verb. Although, this specific phrase seems like a comment rather than a docstring.

webknjaz commented on piptools/utils.py:
```diff
@@ -525,11 +522,11 @@ def override_defaults_from_config_file(
     file, returning the ``pathlib.Path`` of that config file if specified or
     discovered.
 
-    ``None`` is returned if no such file is found.
```
The original phrase was fine since it's not the title line and is in the middle of the prose / details.

chrysle commented:
Addressed your suggestions, thanks! I also moved the `tox` session change into another PR  (#2061).

> Things like this are linkable

I think the [`:py:data:`](https://www.sphinx-doc.org/en/master/usage/domains/python.html#role-py-data) role is only suited for module-level variables? At least Sphinx gives my a corresponding error.

webknjaz commented:
> I think the [`:py:data:`](https://www.sphinx-doc.org/en/master/usage/domains/python.html#role-py-data) role is only suited for module-level variables? At least Sphinx gives my a corresponding error.

The role is correct: https://webknjaz.github.io/intersphinx-untangled/docs.python.org/. The problem is that intersphinx isn't integrated (which I didn't realize), meaning that Sphinx can't look up any foreign objects from the CPython docs.

chrysle commented:
>  The problem is that intersphinx isn't integrated (which I didn't realize), meaning that Sphinx can't look up any foreign objects from the CPython docs.

I didn't think of that, too. Now done! Strangely GH doesn't update this PR after my force-push to this branch?

webknjaz commented:
> ~Strangely GH doesn't update this PR after my force-push to this branch?~ Ah, looks like it works, now.

Sometimes, GitHub is drunk and messes up internal state due to its eventual consistency. It's happening around their outages, usually.

It's good it caught up with the changes. Sometimes, it doesn't and a PR may remain in a weird state forever. Such cases are only fixed with a new PR.

webknjaz commented on docs/conf.py:
```diff
@@ -36,7 +36,12 @@
 # ones.
-extensions = ["myst_parser", "sphinxcontrib.programoutput"]
+extensions = [
+    "myst_parser",
+    "sphinxcontrib.apidoc",
+    "sphinx.ext.intersphinx",
+    "sphinxcontrib.programoutput",
```
These are imported and it's probably best to sort them following the same principles as normal import statements. I usually sort them as follows:
```suggestion
    # Stdlib extensions:
    "sphinx.ext.intersphinx",

    # Third-party extensions:
    "myst_parser",
    "sphinxcontrib.apidoc",
    "sphinxcontrib.programoutput",
```

webknjaz commented on piptools/resolver.py:
```diff
@@ -194,9 +200,8 @@ def __init__(
-        This class resolves a given set of constraints (a collection of
-        InstallRequirement objects) by consulting the given Repository and the
-        DependencyCache.
+        Make sure the legacy resolver is enabled and no backtracking resolver
```
This new phrasing feels weird. That's not what the initializer seems to do...

chrysle replied:
I have no expertise in writing docstrings for `__init__` functions. What would you expect it to contain? A listing of all the parameters? 

>  That's not what the initializer seems to do...

Well, it _does_ raise `PipToolsError` if the user didn't enable the legacy resolver, and removes an entry for the 2020 resolver from the enabled features?

webknjaz commented on piptools/sync.py:
```diff
@@ -43,13 +43,14 @@
-    """
-    Calculate the dependency tree for the package `root_key` and return
-    a collection of all its dependencies.  Uses a DFS traversal algorithm.
+    """Calculate the dependency tree for a package
```
```suggestion
    """Calculate the dependency tree for a package.
```

chrysle commented:
@webknjaz Addressed your comments. Unfortunately, the time-consumingly documented `__init__` method of the `LegacyResolver` class doesn't show up in the built documentation. Do you think that could be related to sphinx-doc/sphinx#11181?

And, I was forced to add `pip`'s `IndexContent` class to `nitpick_ignore_regex`, as Sphinx complained about a missing reference target for the same due to this inheritance:

https://github.com/jazzband/pip-tools/blob/347fec59b418b6239e47dee8429e8401ac4330e6/piptools/_compat/pip_compat.py#L67

I tried setting [`autodoc_inherit_docstrings`](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_inherit_docstrings) to `False` in `conf.py`, but it hasn't helped.

chrysle commented:
@webknjaz I really screwed up the rebasing. Now there are several redundant commits. Do you have any suggestion how to get rid of them?

webknjaz commented:
@chrysle interactive rebase is your friend. And remember to rebase on top of upstream, not your fork. If you can't figure it out, let me know which commits you want to keep and I'll do it myself when I get to PyCon.

chrysle commented:
@webknjaz?

webknjaz commented on piptools/utils.py:
```diff
@@ -266,14 +266,13 @@ def lookup_table_from_tuples(values: Iterable[tuple[_KT, _VT]]) -> dict[_KT, set
 def dedup(iterable: Iterable[_T]) -> Iterable[_T]:
-    """Deduplicate an iterable object like iter(set(iterable)) but
+    """
+    Deduplicate an iterable object like iter(set(iterable)) but
```
```suggestion
    Deduplicate an iterable object like ``iter(set(iterable))`` but
```

webknjaz commented on piptools/utils.py:
```diff
@@ -543,11 +540,11 @@ def override_defaults_from_config_file(
     discovered.
 
-    ``None`` is returned if no such file is found.
+    :returns: :py:data:`None` if no such file is found.
```
```suggestion
    :returns: :py:data:`None` if no such file is found, else returns the path.
```
